helix-term/src/commands.rs

```diff
@@ -5879,39 +5879,70 @@ fn record_macro(cx: &mut Context) {
         keys.pop();
         let s = keys
             .into_iter()
-            .map(|key| format!("{}", key))
-            .collect::<Vec<_>>()
-            .join(" ");
+            .map(|key| {
+                let s = key.to_string();
+                if s.chars().count() == 1 {
+                    s
+                } else {
+                    format!("<{}>", s)
+                }
+            })
+            .collect::<String>();
         cx.editor.registers.get_mut(reg).write(vec![s]);
         cx.editor
-            .set_status(format!("Recorded to register {}", reg));
+            .set_status(format!("Recorded to register [{}]", reg));
     } else {
         let reg = cx.register.take().unwrap_or('@');
         cx.editor.macro_recording = Some((reg, Vec::new()));
         cx.editor
-            .set_status(format!("Recording to register {}", reg));
+            .set_status(format!("Recording to register [{}]", reg));
     }
 }
 
-fn play_macro(cx: &mut Context) {
+fn replay_macro(cx: &mut Context) {
     let reg = cx.register.unwrap_or('@');
-    let keys = match cx
-        .editor
-        .registers
-        .get(reg)
-        .and_then(|reg| reg.read().get(0))
-        .context("Register empty")
-        .and_then(|s| {
-            s.split_whitespace()
-                .map(str::parse::<KeyEvent>)
-                .collect::<Result<Vec<_>, _>>()
-                .context("Failed to parse macro")
-        }) {
-        Ok(keys) => keys,
-        Err(e) => {
-            cx.editor.set_error(format!("{}", e));
-            return;
+    // TODO: macro keys should be parsed one by one and not space delimited (see kak)
+    let keys: Vec<KeyEvent> = if let Some([keys_str]) = cx.editor.registers.read(reg) {
+        let mut keys_res: anyhow::Result<_> = Ok(Vec::new());
+        let mut i = 0;
+        while let Ok(keys) = &mut keys_res {
+            if i >= keys_str.len() {
+                break;
+            }
+            if !keys_str.is_char_boundary(i) {
+                i += 1;
+                continue;
+            }
+
+            let s = &keys_str[i..];
+            let mut end_i = 1;
+            while !s.is_char_boundary(end_i) {
+                end_i += 1;
+            }
+            let c = &s[..end_i];
+            if c != "<" {
+                keys.push(c);
+                i += end_i;
+            } else {
+                match s.find('>').context("'>' expected") {
+                    Ok(end_i) => {
+                        keys.push(&s[1..end_i]);
+                        i += end_i + 1;
+                    }
+                    Err(err) => keys_res = Err(err),
+                }
+            }
```

Comment:
@cessen Can you please help to see if this is correct or if there is a better way to do this?

Reply:
Unless I'm misunderstanding something, I think(?) this is fine.  Since we're just processing key strokes for macros, not actual text data per se, working with char boundaries should be fine.

Although I have no idea how this would interact with IME.  Unfortunately, that's a bit outside my area of expertise, so I don't have any real advice there. :-/

Reply:
I tried to make it function with multibyte characters since if such characters are typed in insert mode while recording, they will end up in the recorded macro. Will this not work if there are multibyte characters?